### ISSUE TITLE
Password Reset Flow not working as documented

### ISSUE THREAD

**Issue by djemrose** [supabase/gotrue-js#349]
# Bug report

## Describe the bug

As heavily discussed there is a bug that logs the user in when they go to reset their password: https://github.com/supabase/supabase/discussions/3360

The `onAuthStateChange` shows this happening by firing the `SIGNED IN` event before the `PASSWORD_RECOVERY` event. I would imagine this is undesirable for everyone. This also causes the hash-bang url to get cleared in the browser immediately on page load, which prevents some apps from grabbing the documented `access_token` for later use.

## To Reproduce

1. Follow the official guide here whilst listening to the `onAuthStateChange` event: https://supabase.com/docs/reference/javascript/auth-api-resetpasswordforemail
2. Also witness the hash-bang url being cleared on page load.

## Expected behavior

1. The hashbang url to not be cleared
2. Only the PASSWORD_RECOVERY event to fire.
3. No SIGNED_IN event should fire on page load.

## Screenshots

![Screen Shot 2022-02-10 at 8 56 21 pm](https://user-images.githubusercontent.com/363101/153383318-fce08269-bce4-4ee3-aa3e-9cf6fc326aa9.png)

## System information

- OS: macOS
- Browser Chrome
- Version of supabase-js: @supabase/supabase-js 1.1.2
- Version of Node.js: v16.13.1

**Comment by funwithtriangles** [comment]
Having a similar issue, I just get `SIGNED_IN` instead of `PASSWORD_RECOVERY`

**Comment by funwithtriangles** [comment]
FWIW I fixed my issue by avoiding the auth event altogether and just making sure the link redirects the user to the correct page:

```ts
await supabase.auth.resetPasswordForEmail(email, {
    redirectTo: `${window.location.origin}/reset-password`,
  })
```

**Comment by hf** [comment]
Please avoid relying on the sequence of events fired in the `onAuthStateChange` callback as the order is never guaranteed.

You can always use [`getSession()`](https://supabase.github.io/gotrue-js/v2/classes/GoTrueClient.html#getSession) to safely access the latest access token.

**Comment by laygir** [comment]
I'm quite new to Supabase, I've been hearing it a lot and wanted to give it a go in a side project. 
But seeing this issue open for more than a year is a bit surprising considering its impact, but hopefully the fix is on the way I believe 🤞  #629 

Meanwhile, before I came across #629 and after spending several hours trying to make my frontend flow nice and tight, I gave up on using `onAuthStateChange` and I am not sure what I am losing by not using it.

Fwiw, to prevent user from being logged it white reset password op, I do this;

1. I am using the following in my Reset Password email:
```html
<p><a href="{{ .SiteURL }}/reset-password?confirmation_url={{ .ConfirmationURL }}%26token%3d{{ .Token}}%26email%3d{{ .Email}}">Reset Password
```
2. Then when user lands on `/reset-password` without doing any `auth.getSession()`, pick up the `confirmation_url` from url, run it through `new URLSearchParams`, extract the `token`, `email`, `type`.

3. Manually call `auth.verifyOtp({ token, email, type })`

4. Finally call `updatePassword({ password })`

**Comment by abdulrahimiliasu** [comment]
Any update on this ?

**Comment by kangmingtay** [comment]
Hey @abdulrahimiliasu and @laygir, we've already merged in the PR to fix this issue: https://github.com/supabase/gotrue-js/pull/629 so we'll be closing this issue. Please feel free to reopen it if you're still experiencing the same issue.

**Comment by jonathanlal** [comment]
yeah not getting PASSWORD_RECOVERY event either...

**Comment by nookitapp** [comment]
@jonathanlal We're you able to fix? Running into the same issue despite the fix being released.

**Comment by JeremyMees** [comment]
I'm also still having this issue

**Comment by m-sadegh-sh** [comment]
I'm facing the same issue (supabase-js 2.39.8)

**Comment by knielsen24** [comment]
After clicking confirmation (redirect) url, I am NOT getting any event to fire in my session provider.  My redirect takes me to the correct path, but if I submit the updateUser method with the new password, I get an error response "Auth session missing!".  Other events, initial session, signed in, signed out are working just fine.

supabase-js: 2.40.0

I'm using React Native bare project with expo modules.

**Comment by kimhwanhoon** [comment]
I was having the same problem too. I never saw `PASSWORD_RECOVERY`.
But then I was thinking that it might be bc of the "Email Templates" on supabase. Since at first, `confirm signup template` and `reset password template` both use `<a href="{{ .ConfirmationURL }}"</a>`, maybe we should add some additional things for our app to recognize the emit `PASSWORD_RECOVERY`. I’ll be working on it and come back if I get anything

**Comment by Chizaruu** [comment]
> I'm quite new to Supabase, I've been hearing it a lot and wanted to give it a go in a side project. But seeing this issue open for more than a year is a bit surprising considering its impact, but hopefully the fix is on the way I believe 🤞 #629
> 
> Meanwhile, before I came across #629 and after spending several hours trying to make my frontend flow nice and tight, I gave up on using `onAuthStateChange` and I am not sure what I am losing by not using it.
> 
> Fwiw, to prevent user from being logged in while reset password op, I do this;
> 
>     1. I am using the following in my Reset Password email:
> 
> 
> ```
> <p><a href="{{ .SiteURL }}/reset-password?confirmation_url={{ .ConfirmationURL }}%26token%3d{{ .Token}}%26email%3d{{ .Email}}">Reset Password
> ```
> 
>     2. Then when user lands on `/reset-password` without doing any `auth.getSession()`, pick up the `confirmation_url` from url, run it through `new URLSearchParams`, extract the `token`, `email`, `type`.
> 
>     3. Manually call `auth.verifyOtp({ token, email, type })`
> 
>     4. Finally call `updatePassword({ password })`

Thanks for the help! I had to change it up a bit to verify OTP. (it must have changed after one year)

`const** { error } = await supabase.auth.verifyOtp({ type, token_hash: token });`

For all the SvelteKit lurkers, I'll post a repo at some point and edit this comment.

**Comment by Staff2Stand** [comment]
> I'm quite new to Supabase, I've been hearing it a lot and wanted to give it a go in a side project. But seeing this issue open for more than a year is a bit surprising considering its impact, but hopefully the fix is on the way I believe 🤞 #629
> 
> Meanwhile, before I came across #629 and after spending several hours trying to make my frontend flow nice and tight, I gave up on using `onAuthStateChange` and I am not sure what I am losing by not using it.
> 
> Fwiw, to prevent user from being logged in while reset password op, I do this;
> 
> 1. I am using the following in my Reset Password email:
> 
> ```
> <p><a href="{{ .SiteURL }}/reset-password?confirmation_url={{ .ConfirmationURL }}%26token%3d{{ .Token}}%26email%3d{{ .Email}}">Reset Password
> ```
> 
> 2. Then when user lands on `/reset-password` without doing any `auth.getSession()`, pick up the `confirmation_url` from url, run it through `new URLSearchParams`, extract the `token`, `email`, `type`.
> 3. Manually call `auth.verifyOtp({ token, email, type })`
> 4. Finally call `updatePassword({ password })`

I was having the same issue and this worked for me!
